### PR TITLE
Betareg rebased3 Beta regression

### PR DESCRIPTION
**Pull request by josef-pkt**
rebased version of #7282 Betareg_rebased2

This PR includes changes to the GenericLikelihoodModel classes

earlier PRs superseded by this PR
closes #7282 ,  closes #4238 ,  closes #4238 #2030

API still needs improvements

### CODE REVIEW

**Comment by josef-pkt** [comment]
If there are no serious problems here, then I can continue with the previous `rebased2` branch and PR (that one is only 5 months old)

File: statsmodels/miscmodels/betareg.py [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+
+            jbg = phi * t * h * (mu * var_star + c)
+            if observed:
+                jbg -= ymu_star * t * h
+
+            jgg = h**2 * (mu**2 * var_star + 2 * mu * c + var_t)
+            if observed:
+                jgg += (mu * ymu_star + yt - mut) * q * h**3    # **3 ?
+
+        return np.column_stack((sf1, sf2)), (-jbb, -jbg, -jgg)
```
Comment (josef-pkt) [inline statsmodels/miscmodels/betareg.py:312]: has return_hessian
but seems to return it always

File: statsmodels/miscmodels/tests/test_beta.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+income = pd.read_csv(op.join(HERE, 'foodexpenditure.csv'))
+methylation = pd.read_csv(op.join(HERE, 'methylation-test.csv'))
+
+def check_same(a, b, eps, name):
+    assert np.allclose(a, b, rtol=0.01, atol=eps), \
+            ("different from expected", name, list(a), list(b))
+
+def assert_close(a, b, eps):
+    assert np.allclose(a, b, rtol=0.01, atol=eps), (list(a), list(b))
```
Comment (josef-pkt) [inline statsmodels/miscmodels/tests/test_beta.py:54]: should be obsolete
use assert_allclose directly

File: statsmodels/miscmodels/betareg.py [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+        """
+        mean = self.predict(params, exog=exog)
+        precision = self.predict_precision(params,
+                                           exog_precision=exog_precision)
+
+        var_endog = mean * (1 - mean) / (1 + precision)
+        return var_endog
+
+
+    def nloglikeobs(self, params):
```
Comment (josef-pkt) [inline statsmodels/miscmodels/betareg.py:172]: I will deprecate nloglikeobs
just use loglikeobs

**Comment by josef-pkt** [comment]
`@josef-pkt` in commit message sends an email to me
interactive rebase to edit ?

**Comment by josef-pkt** [comment]
`get_distribution` is a results method and does not have `predict` options.

**Comment by josef-pkt** [comment]
for unit tests we need additional statistics like llf, aic,... from R package

**Comment by josef-pkt** [comment]
many pep-8 failures, whitespace, line length, imports

**Comment by josef-pkt** [comment]
test coverage looks good, missing mainly score_hessian_factor and score_obs
tolerance, precision required in unit tests against R is very loose.

**Comment by josef-pkt** [comment]
appveyor only fails in style check and unrelated `pre` testing

File: statsmodels/miscmodels/betareg.py [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+        Zparams = params[-nz:]
+
+        mu = self.link.inverse(np.dot(X, Xparams))
+        phi = self.link_precision.inverse(np.dot(Z, Zparams))
+
+        alpha = mu * phi
+        beta = (1 - mu) * phi
+
+        if np.any(alpha <= np.finfo(float).eps): return np.array(-np.inf)
+        if np.any(beta <= np.finfo(float).eps): return np.array(-np.inf)
```
Comment (josef-pkt) [inline statsmodels/miscmodels/betareg.py:197]: these return scalar instead of array.
remove or refactor

**Comment by josef-pkt** [comment]
HC0 cov_type works without raising exception
`res = mod.fit(method="bfgs", cov_type="HC0")`

**Comment by josef-pkt** [comment]
maybe precision problems in hessian
1.83953039 versus R: 1.79098  for bse of precision constant (default log link)

currently no analytical hessian.
score_hessian_factor method is there but not used

```
print(ss_meth)

_methylation_estimates_mean = u"""varname      Estimate StdError zvalue Pr(>|z|)
(Intercept)  1.44224    0.03401  42.404   2e-16
genderM      0.06986    0.04359   1.603    0.109
CpGCpG_1     0.60735    0.04834  12.563   2e-16
CpGCpG_2     0.97355    0.05311  18.331   2e-16"""

_methylation_estimates_precision = u"""varname Estimate StdError zvalue Pr(>|z|)
(Intercept)  8.22829    1.79098   4.594 4.34e-06
age         -0.03471    0.03276  -1.059    0.289"""

res_meth.params
array([ 1.44224318,  0.06985727,  0.60734527,  0.9735476 ,  8.22828403,
       -0.03470541])
res_meth.bse
array([0.03404539, 0.04358708, 0.04900909, 0.05338694, 1.83953039,
       0.03364145])
```

**Comment by josef-pkt** [comment]
looking good: R uses EIM
we have `hessian_1` based on analytical hessian factor with `observed` option
analytical hessian with default  observed=True is close to numerical hessian

using EIM, the bse for precision parameters are close to R's

```
eim = res_meth.model.hessian_1(res_meth.params, observed=False)
np.sqrt(np.diag(np.linalg.inv(-eim)))
array([0.03401209, 0.04359128, 0.0483425 , 0.05311043, 1.79098056,
       0.03276481])

in R
(Intercept)  8.22829    1.79098   4.594 4.34e-06
age         -0.03471    0.03276  -1.059    0.289"""

```

**Comment by josef-pkt** [comment]
currently df_xxx only take mean slope parameters into account

meth example: 6 parameters: 3 mean slope, 1 mean intercept, 1 precision intercept, 1 precision slope

```
res_meth.df_resid, res_meth.df_model, res_meth.model.nparams, res_meth.nobs, res_meth.nobs - res_meth.model.nparams
(32.0, 3.0, 6, 36, 30)
```

aic uses 4 params, but should use 6 params, precision params are missing

```
res_meth.llf, res_meth.aic, 2 * (res_meth.llf - 4), 2 * (res_meth.llf - 6)
(array(104.14802841),
 array(-200.29605681),
 array(200.29605681),
 array(196.29605681))
```

**Comment by josef-pkt** [comment]
this might just be a few days of work to get the basic version merged (as long as I don't start to read articles)

`get_influence` should work generically for MLEInfluence, but I doubt I can write unit tests against other packages. The articles might have replicable results. (I had read articles for beta regression, among others, when writing MLEInfluence, AFAIR)

predict:
`predict` only has mean prediction, no "which" option, no linpred
two extra predict methods for precision and variance, predict_precision and predict_variance    (those are non-standard in statsmodels)

still no `llnull`

summary has just the basics, e.g.

```
                                 Beta Results                                 
==============================================================================
Dep. Variable:            methylation   Log-Likelihood:                 104.13
Model:                           Beta   AIC:                            -200.3
Method:            Maximum Likelihood   BIC:                            -193.9
Date:                Tue, 06 Jul 2021                                         
Time:                        14:19:54                                         
No. Observations:                  36                                         
Df Residuals:                      32                                         
Df Model:                           3                                         
================================================================================
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        1.4422      0.041     35.104      0.000       1.362       1.523
gender[T.M]      0.0721      0.047      1.530      0.126      -0.020       0.164
CpG[T.CpG_1]     0.6051      0.051     11.777      0.000       0.504       0.706
CpG[T.CpG_2]     0.9746      0.051     18.927      0.000       0.874       1.076
precision-1   1693.2403    944.263      1.793      0.073    -157.480    3543.961
precision-2    -20.4039     16.526     -1.235      0.217     -52.793      11.986
================================================================================
```

**Comment by josef-pkt** [comment]
R betareg has several types of residuals, with reference
`residuals(object, type = c("sweighted2", "deviance", "pearson", "response", "weighted", "sweighted") ...`

**Comment by josef-pkt** [comment]
I don't find a reference to methylation data, might be from applications that Brent Pederson had

**Comment by josef-pkt** [comment]
Based on the results attributes, R betareg does not define df_model, but it has df_null
In this two part model (mean, precision) we need two constant in the null model.

```
food_id.n = 38
food_id.nobs = 38
food_id.df_null = 36
food_id.df_residual = 34
```

Aside: what if null model is not nested in full model, i.e. no constant?
LR test will not be for nested models.
No constant will be very unlikely in this case.

---

getting standard errors `bse` out of R:
bse are part of `coefficients`. That includes also z-value and p-value, no confint in R summary

```
> coefficients(summ)$mean[, 1:2]
                        Estimate           Std. Error
(Intercept) -0.62254805618912845 0.223853539305630178
income      -0.01229884053492592 0.003035584648662564
persons      0.11846209769288003 0.035340667012308098

> summ$coefficients$mean[, 1:2]
                        Estimate           Std. Error
(Intercept) -0.62254805618912845 0.223853539305630178
income      -0.01229884053492592 0.003035584648662564
persons      0.11846209769288003 0.035340667012308098

> summ$coefficients$precision[, 1:2]
          Estimate         Std. Error 
35.609750328915382  8.079598247958925
```

**Comment by josef-pkt** [comment]
RuntimeWarning in unit tests if mu=0 in example, 
This happens during optimization, so away from MLE.


```
statsmodels/miscmodels/tests/test_beta.py::TestBeta::test_income_coefficients
  ...\statsmodels\statsmod
els\miscmodels\betareg.py:254: RuntimeWarning: invalid value encountered in subt
ract
    sf2 = h * (mu * (ystar - mustar) + yt - mut)
```

nan in sf2 because of `inf - inf`

```
        sf2 = h * (mu * (ystar - mustar) + yt - mut)
        if not np.all(np.isfinite(sf2)):
>           raise
E           RuntimeError: No active exception to reraise

...\statsmodels\miscmodels\betareg.py:256: RuntimeError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>> PDB post_mortem (IO-capturing turned off) >>>>>>>>>>>>>>>>>>
> ...\statsmodels\statsmodels\miscmodels\betareg.py(256)score_factor()
-> raise
(Pdb) sf2
array([          nan,           nan,           nan,           nan,
                 nan,  -99.63639881,           nan,           nan,
                 nan,           nan,           nan,           nan,
                 nan,           nan,           nan, -100.9011225 ,
                 nan, -117.06403449,           nan,           nan,
                 nan,           nan,           nan,           nan,
                 nan,           nan,           nan,           nan,
                 nan,  -57.00328273,           nan,           nan,
                 nan,           nan,           nan,           nan,
                 nan,  -60.32615123])
(Pdb) mut
array([           -inf,            -inf,            -inf,            -inf,
                  -inf, -4.48472176e+13,            -inf,            -inf,
                  -inf,            -inf,            -inf,            -inf,
                  -inf,            -inf,            -inf, -2.56504333e+09,
                  -inf, -3.98641934e+10,            -inf,            -inf,
                  -inf,            -inf,            -inf,            -inf,
                  -inf,            -inf,            -inf,            -inf,
                  -inf, -1.89942051e+09,            -inf,            -inf,
                  -inf,            -inf,            -inf,            -inf,
                  -inf, -9.46143831e+09])
(Pdb) mu
array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1.])
(Pdb) digamma(0)
-inf
```

```
(Pdb) digamma(1e-20)
-1e+20
```

**Comment by josef-pkt** [comment]
no wrapper class yet for the results
my notebook doesn't have pandas and formula example, except for data handling outside model

**Comment by josef-pkt** [comment]
`subset` option in `from_formula` is not unit tested.
I'm not sure we really should support that option. I never even used it. It's easier just to do it outside the model with pandas or numpy.

**Comment by josef-pkt** [comment]
most basic things are done

- predict which including `precision`
- get_distribution for out of sample exog

Stata predict has:
```
cmean conditional mean of the dependent variable; the default
cvariance conditional variance of the dependent variable
xb linear prediction in the conditional-mean equation
xbscale linear prediction in the conditional-scale equation
stdp standard error of the linear prediction
```

Can we get `get_prediction` standard errors and confint for predicted mean (indep of variance/precision) ?
Then we could use something similar to generic get_prediction using subser of params`

Stata also has specific `margins`, all other post-estimation results are generic (no specific documentation)

Stata 
`vce(vcetype):  vcetype may be oim, robust, cluster clustvar, bootstrap,or jackknife`

`fweights, iweights, and pweights are allowed;`
this does not include aweights, i.e. var_weights.

linear constraints are allowed but just links to generic documentation. I don't know what the cross-parameter, mean scale, restrictions for this are.

**Comment by josef-pkt** [comment]
missing keys for recreating model and remove_data

llnull:
DiscreteResults.llnull has a general version recreating a auxiliary constant-only model.
ZeroInflated models don't have llnull specifically defined. maybe generic works also for two part models.

`GenericZeroInflated.__init__` has

```
        self._init_keys.extend(['exog_infl', 'inflation'])
        self._null_drop_keys = ['exog_infl']
```

how does offset and exposure get into `_init_keys`? I don't see it in discrete.discrete_model

**Comment by josef-pkt** [comment]
see #7569 

after I add kwds to the super call in `GenericLikelihoodModel.__init__`, I get

```
init_keys
mod_meth._init_keys
['exog_precision', 'missing_idx', 'formula', 'design_info']
```

for model created with 
`mod_meth = Beta.from_formula(model, methylation, exog_precision=Z)`

Why does this include `'missing_idx', 'formula', 'design_info'`?
They wouldn't apply for llnull model, or other models with transformed exog

**Comment by josef-pkt** [comment]
OIM in R betareg

`betareg.control` has a hessian option
`hessian logical. Should the numerical Hessian matrix from the optim output be used  
for estimation of the covariance matrix?`

bfgs numerical hessian might not be very precise

```
>res_meth = betareg(methylation ~ gender + CpG | age, data=meth, hessian=TRUE)
> summary(res_meth)

In this case, bse for precision params in R agree at 3 decimals with our analytical oim hessian

**Comment by josef-pkt** [comment]
cov_type="eim" to be able to match R betareg

I'm cheating for now and use a model attribute `hess_type` to switch between observed is True or False
```
        if "cov_type" in kwds:
            # this is a workaround because we cannot tell super to use eim
            if kwds["cov_type"].lower() == "eim":
                self.hess_type = "eim"
                del kwds["cov_type"]
            else:
                self.hess_type = "oim"
```

**Comment by josef-pkt** [comment]
R betareg: pseudo rsquared looks strange

```
pseudo.r.squared
pseudo R-squared value (squared correlation of linear predictor and link-transformed response),
```

we use McFadden's pseudo rsquared, based on likelihood ratio to null model
(but at the moment I have prsquared negative, code copied from DiscreteResults ???)

R betareg has df_null in results, but I don't see llnull or llr, llr_pvalue for likelihood ratio compared to null

**Comment by josef-pkt** [comment]
what I have now for prsquared:

base.model has a new Mixin `_LLRMixin` for likelihood ratio, llnull, and a method
`pseudo_rsquared(self, kind="mcf")`
that includes McFadden and Cox-Snell/Maddala/likelihood ratio rsquared.

cox-snell is from #7367 PR for GLM

In BetaResults, I added the cached attribute `prsquared` based on cox-snell likelihood ratio 

pseudo_rsquared(self, kind="mcf")

**Comment by josef-pkt** [comment]
next commit will be moving Beta regression to new `statsmodels.othermod`

I want a more generic module name for `interval` data/endog. e.g. kuramaswami or others might follow

I'm going to add full unit tests for what's available in R betareg.
I would need Stata results for llnull and related, or just test against an user specified null model (my current check in notebook)

still missing: offsets, pickle/remove_data support

**Comment by josef-pkt** [comment]
the latest changes break  the TModel test `statsmodels.miscmodels.tests.test_tmodel.TestTModelFixed` for missing values
I guess this will be related to adding `**kwargs` to the super call in `GenericLikelihoodModel.__init__`

**Comment by josef-pkt** [comment]
running the unit tests for TModel locally, I get a warning

```
statsmodels/miscmodels/tests/test_tmodel.py::TestTModelFixed::test_smoke
  ...\statsmodels\statsmod
els\base\model.py:2630: UserWarning: df_resid differs from nobs - nparams
    warnings.warn("df_resid differs from nobs - nparams")
```

need to check how df are defined with fixed df, maybe there is something wrong
Could be related to recent changes in GenericLikelihood if TModel doesn't have unit tests for `df_resid`

**Comment by josef-pkt** [comment]
This fails for many cases in test_ar

```
>   combined.update({k: kwargs[k] for k in none_array_names})
E   KeyError: 'exog'
```

This doesn't handle exog=None correctly, because exog is not a kwarg, and is explicitly added to none_array_names

```
            if exog is None:
                none_array_names += ['exog']
```


The case in `if` handles both cases with None value and cases with non-array values like scalars
endog, exog are the only args, besides separate missing

**Comment by josef-pkt** [comment]
going to be all green beside a few style check failures

time to move to `othermod`

**Comment by josef-pkt** [comment]
all green, including style check
`pre` testing still fails in statespace

get_distribution
I extended the results class version. But that does not take `params`, so it's less useful for simulation for, e.g., power where we have user specified params.

But if I use model.predict, then we don't get the transform/formula handling from inherited Results.predict.
I think that's also a limitation of current GLM.get_distribution.

Also inherited Results.predict cannot handle formula for second exog, i.e. exog_precision_formula in current api.

**Comment by josef-pkt** [comment]
problem in predict with auxiliary data arrays

currently it assumes either both exog are specified or neither.
If only exog is given as arg not None, and exog_precision is None, then we should not take the model exog_precision.

GenericZeroInflated does not check for a default exog_infl, so user will have to specify all or nothing.
Similar to offset, exposure in GLM, we can have the case where exog_precision is constant, and we use a defaulte if exog_precision is None. In that case predict needs to check whether exog_precision is constant-only.

**Comment by josef-pkt** [comment]
changing get_distribution to go through predict
predict inherits formula transformation for `exog`, but it's not available for `exog_precision`

two separate get_distribution with currently no code sharing, 
- results.get_distribution includes transformation from results.predict
- model.get_distribution uses model.predict and needs numpy arrays

**Comment by josef-pkt** [comment]
results name BetaRegressionResults doesn't match the pattern, need to rename to ???

**Comment by josef-pkt** [comment]
all green, merging
